src/merkle/merkle_tree.rs

```diff
@@ -19,31 +20,84 @@ impl MerkleTree {
     ///
     /// # Errors
     /// Returns an error if the number of leaves is smaller than two or is not a power of two.
-    pub fn new(leaves: Vec<Word>) -> Result<Self, MerkleError> {
+    pub fn new<T, I, V>(leaves: T) -> Result<Self, MerkleError>
+    where
+        T: IntoIterator<IntoIter = I>,
+        I: ExactSizeIterator + Iterator<Item = V>,
+        V: Borrow<Word>,
+    {
+        let leaves = leaves.into_iter();
         let n = leaves.len();
         if n <= 1 {
             return Err(MerkleError::DepthTooSmall(n as u8));
         } else if !n.is_power_of_two() {
             return Err(MerkleError::NumLeavesNotPowerOfTwo(n));
         }
 
-        // create un-initialized vector to hold all tree nodes
-        let mut nodes = unsafe { uninit_vector(2 * n) };
-        nodes[0] = [Felt::ZERO; 4];
-
-        // copy leaves into the second part of the nodes vector
-        nodes[n..].copy_from_slice(&leaves);
-
-        // re-interpret nodes as an array of two nodes fused together
-        // Safety: `nodes` will never move here as it is not bound to an external lifetime (i.e.
-        // `self`).
-        let ptr = nodes.as_ptr() as *const [RpoDigest; 2];
-        let pairs = unsafe { slice::from_raw_parts(ptr, n) };
+        // The tree contains `n` leaves, which is a power of two. The depth of the tree is `d`
+        // defined as `2^d=n` or `d=lb(n)`. The total number of elements in the tree is
+        // `2**(d+1)-1` or `2n-1` (since `n=2**d`).
+        //
+        // Below we compute the tree size `+1`, the first element is set to `0`, so that the length
+        // of `nodes` will be a power of two.
+        let final_capacity = 2 * n;
+
+        // Allocate data to accomodate all the elements, and get a reference to the underlying
+        // buffer
+        let mut nodes: Vec<Word> = Vec::with_capacity(final_capacity);
+        let buffer = nodes.spare_capacity_mut();
```

Comment:
I don't think this is an isolated problem to this case. We have `uninit_vector` in other places too. I think we should:

1) Decide what is the best design for such function
2) Generalize it somewhere (either `winter-utils` or `miden-utils`)
3) Replicate the behavior

Reply:
I agree with you, but I wanted to do one change at the time (doing this change took quite a bit of time with reading the code / docs).

Once we agree on what are the issues, and the strategy to fix it. I can do implement the changes on the other places too :)